plugin/pkg/admission/serviceaccount/admission.go

```diff
@@ -282,6 +285,51 @@ func (s *serviceAccount) enforceMountableSecrets(serviceAccount *corev1.ServiceA
 	return false
 }
 
+func (s *serviceAccount) mountVolumeForTokenRequest(pod *api.Pod, a admission.Attributes) {
+	apiVersion := a.GetResource().Version
+	if apiVersion == "" {
+		apiVersion = "v1"
+	}
+	for _, vol := range pod.Spec.Volumes {
+		if vol.Projected != nil {
+			for _, pro := range vol.Projected.Sources {
+				if pro.ServiceAccountToken != nil {
```

Comment:
this seems to be locating and appending ca and namespace to an existing service account token projected volume, which is not what I expected. there are use cases for injecting a token without the namespace and CA.

I would only expect us to inject namespace and CA for the volume we are setting up to replace the existing service account token (the one we set up inside `mountServiceAccountToken`)

Reply:
@liggitt  We can't mount two volumes to the same mount path. We need to make a choice here:

1. Create a independent volume which contains namespace&ca.crt projection volume, then mount it to the sub path of the path which token was mounted. 

2.  Create a independent volume which contains namespace and ca,crt projection volume, then 
mount it to a fix path like "/var/run/secrets/kubernetes.io/". 

3. Append ca.crt and namespace projection volume to the existing service account token projected volume just as my patch did.

I think for option 1&2, it will break current situation that ca.crt, namespace, token file appear in same path. And it is very hard to do so if user want. So maybe option3 is better choice. For the concern you proposed, if user  doesn't need namespace and CA, we just give up appending ca&namespace projection volume.
WDYT

Reply:
> Append ca.crt and namespace projection volume to the existing service account token projected volume just as my patch did

This admission plugin is responsible for creating the volume to be mounted at `/var/run/secrets/kubernetes.io/serviceaccount`. It should either create a volume referencing a service account token secret (like it does today), or (when the projection feature is on and it detects the apiserver is serving tokenrequest endpoints) it should create a projection volume containing the service account token, ca.crt and namespace projections. It should not modify pre-existing projected volumes it finds.

As an optimization, if there is a pre-existing projection volume that perfectly matches the shape of the volume it would add, it can reuse it instead of appending a new volume (just like the code in master does today if there is an existing secret volume referencing the service account token secret).